packages/malloy-render/src/component/render-result-metadata.ts

```diff
@@ -21,283 +22,1562 @@
+      return new DateField(field, parent);
+    } else if (isTimestampFieldInfo(field)) {
+      return new TimestampField(field, parent);
+    } else if (isStringFieldInfo(field)) {
+      return new StringField(field, parent);
+    } else if (isNumberFieldInfo(field)) {
+      return new NumberField(field, parent);
+    } else if (isSQLNativeFieldInfo(field)) {
+      return new SQLNativeField(field, parent);
+    } else {
+      throw new Error(`Unknown field type ${field.type.kind}`);
+    }
+  },
+  isNestField(field: Field): field is NestField {
+    return (
+      field instanceof RepeatedRecordField ||
+      field instanceof ArrayField ||
+      field instanceof RecordField
+    );
+  },
+  pathFromString(path: string) {
+    return JSON.parse(path);
+  },
+  pathToString(path: string[]) {
+    return JSON.stringify(path);
+  },
+};
+
+function tagFor(field: Malloy.DimensionInfo, prefix = '# ') {
+  return tagFromAnnotations(field.annotations, prefix);
+}
+
+const RENDER_TAG_LIST = [
+  'link',
+  'image',
+  'cell',
+  'list',
+  'list_detail',
+  'bar_chart',
+  'line_chart',
+  'dashboard',
+  'scatter_chart',
+  'shape_map',
+  'segment_map',
+];
+
+const CHART_TAG_LIST = ['bar_chart', 'line_chart'];
+
+type DrillValue = {field: Field; value: Cell};
+
+export function shouldRenderChartAs(tag: Tag) {
+  const tagNamesInOrder = Object.keys(tag.properties ?? {}).reverse();
+  return tagNamesInOrder.find(name => CHART_TAG_LIST.includes(name));
+}
+
+export function shouldRenderAs(
+  field: Malloy.DimensionInfo,
+  parent: Field | undefined,
+  tagOverride?: Tag
+) {
+  const tag = tagOverride ?? tagFor(field);
+  const tagNamesInOrder = Object.keys(tag.properties ?? {}).reverse();
+  for (const tagName of tagNamesInOrder) {
+    if (RENDER_TAG_LIST.includes(tagName)) {
+      if (['list', 'list_detail'].includes(tagName)) return 'list';
+      if (['bar_chart', 'line_chart'].includes(tagName)) return 'chart';
+      return tagName;
+    }
+  }
+
+  if (field.type.kind === 'record_type' && parent?.renderAs === 'chart') {
+    return 'none';
+  }
+
+  const isNest =
+    field.type.kind === 'array_type' || field.type.kind === 'record_type';
+
+  if (!isNest) return 'cell';
+  return 'table';
+}
+
+export abstract class FieldBase {
+  public readonly tag: Tag;
+  public readonly path: string[];
+  protected readonly metadataTag: Tag;
+  public readonly renderAs: string;
+  public readonly valueSet = new Set<string | number | boolean>();
+  constructor(
+    public readonly field: Malloy.DimensionInfo,
+    public readonly parent: NestField | undefined
+  ) {
+    this.tag = tagFor(this.field);
+    this.metadataTag = tagFor(this.field, '#(malloy) ');
+    this.path = parent
+      ? parent.isArray()
+        ? [...parent.path]
+        : [...parent.path, field.name]
+      : [];
+    this.renderAs = shouldRenderAs(field, parent);
+  }
+
+  populateAllVegaSpecs(_options: GetResultMetadataOptions) {}
```

Comment:
A possible solution to the JSDOM issue is to init the `RootCell` with a generic type `<MetadataT>` and a function which generates/populates that `metadata`. Then we could hide all the `renderer-next` stuff in that function, and the legacy renderer wouldn't need to import any of it?

Reply:
Or just pull it back out into a separate metadata object like before....